File: 02_ComparisonSorts/Quick Sort/QuickSort.java
import java.io.*;
import java.util.*;

class QuickSort {
    /*
     * creating an array of size n
     * initializing the arry with random values
     */
    public int[] initializeArrayWithRandomNumbers(int n) {
        int arr[] = new int[n];
        int i;
        Random rand = new Random();
        for (i = 0; i < n; ++i) {
            arr[i] = rand.nextInt(n);
        }
        return arr;
    }

    // Swaps the values at position1 and position2 in the array parameter and
    // returns the edited array
    public int[] swap(int[] arr, int position1, int position2) {
        int temporary = arr[position1];
        arr[position1] = arr[position2];
        arr[position2] = temporary;
        return arr;
    }// swap()

    public void printArray(int arr[]) {
        int i;
        for (i = 0; i < arr.length; ++i) {
            System.out.print(arr[i] + "\t");
        }
        System.out.println();
    }

    /*
     * This function takes middle element as pivot,
     * places the pivot element at its correct
     * position in sorted array, and places all
     * smaller (smaller than pivot) to left of
     * pivot and all greater elements to right
     * of pivot
     */
    protected int partitionMiddle(int[] arr, int left, int right) {
        int x = arr[left];
        int y = arr[(left + right) / 2];
        int z = arr[right];
        int pivot = (x < y ? (y < z ? y : x < z ? z : x) : (y > z ? y : x > z ? z : x));

        int i = left - 1;
        int j = right + 1;

        while (i < j) {
            while (arr[++i] < pivot) {
                if (i == right)
                    break;
            } // while i loop (inner)
            while (arr[--j] > pivot) {
                if (j == left)
                    break;
            } // while j loop (inner)
            if (i >= j)
                break;
            arr = swap(arr, i, j);
        } // while true loop (outer)
        return j;
    }// partitionMiddle()

    /*
     * The main function that implements QuickSort()
     * arr[] --> Array to be sorted,
     * low --> Starting index,
     * high --> Ending index
     */
    protected int[] quickSort(int[] arr, int left, int right) {
        int mid;
        if (left < right) {
            mid = partitionMiddle(arr, left, right);
            arr = quickSort(arr, left, mid); // sorting the left half
            arr = quickSort(arr, (mid + 1), right); // sorting the right half
        } // if

        return arr;
    }// quickSort()

}

class MainQuick {
    public static void main(String[] args) {
        QuickSort qs = new QuickSort();
        int size = 10;
        int arrAns[];
        int arr[] = qs.initializeArrayWithRandomNumbers(size);
        System.out.println("Unsorted Array");
        qs.printArray(arr);

        // Bubble Sort
        System.out.println("**********Quick Sort**********");
        arrAns = qs.quickSort(arr, 0, size - 1);
        qs.printArray(arrAns);
    }
}

File: 02_ComparisonSorts/Quick Sort/QuickSort.py
import random
 
"""
 creating an array of size n
 initializing the arry with random values
"""
def initializeArrayWithRandomNumbers(n):
    myList = []
    for i in range(n):
        myList.append(random.randint(0,n))
    return myList
	
def partitionMiddle(myList, left, right):
	x = myList[left]
	y = myList[(left + right)//2]
	z = myList[right]
	
	#select pivot point
	if x < y:
		if y < z:
			pivot = y
		elif x < z:
			pivot = z
		else:
			pivot = x
	else:
		if y > z:
			pivot = y
		elif x > z:
			pivot = z
		else:
			pivot = x
	i = left - 1
	j = right + 1
	while i < j:
		j -= 1
		i += 1
		while myList[i] < pivot: #inner while i
			if i == right:
				break
			i += 1

		while myList[j] > pivot: #inner while j
			if j == left:
				break
			j -= 1
		if i >= j:
			break
		myList[i], myList[j] = myList[j], myList[i]
	return j	
 
def quickSort(myList, left, right):
	if left < right:
		mid = partitionMiddle(myList, left, right)
		myList = quickSort(myList, left, mid)
		myList = quickSort(myList, (mid+1), right)
	return myList

size = 10
print("Unsorted Array")
myList = initializeArrayWithRandomNumbers(size)
print(myList)

print("**********Quick Sort**********")
myList = quickSort(myList, 0, (size-1))
print(myList)